### PR TITLE
Add iterator for combinations

### PR DESCRIPTION
**Pull request by thofma**

### CODE REVIEW

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 1]
```diff
@@ -0,0 +1,82 @@
+@doc raw"""
+    combinations(n::Int, k::Int)
+
+Return an iterator over all $k$-combinations of ${1,...,n}$, produced in
+lexicographically ascending order.
+
+See also
```
Comment (lgoettgens) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: there seems to be something missing here

**Review by lgoettgens — COMMENTED** [review]
Could you please add some more extensive tests? You can either get inspiration at https://github.com/Nemocas/AbstractAlgebra.jl/pull/2034/files#diff-9383dd485d4500b18b2bfa2797938c914f4a8df63c205e8ea8ddcfe23a1ad538 or `test/Combinatorics/EnumerativeCombinatorics/compositions.jl`, but I would expect at least the following tests: edge-cases with zero (in particular `combinations(0, 0)`), something with `allunique`, testing the guaranteed iteration order (e.g. hardcoding the result of `combinations(5,3)` which is the smallest example where lex and revlex order differ, and some large example with `issorted`)

**Comment by lgoettgens** [comment]
This implementation seems to be slower than `AbstractAlgebra.combinations`, even when wrapping another `collect` around it.

```julia
julia> @b collect(AbstractAlgebra.combinations(5,3))
615.111 ns (28 allocs: 1.578 KiB)

julia> @b collect(combinations(5,3))
807.188 ns (34 allocs: 1.312 KiB)

julia> @b collect(AbstractAlgebra.combinations(15,10))
250.671 μs (6027 allocs: 547.938 KiB)

julia> @b collect(combinations(15,10))
280.748 μs (9014 allocs: 539.844 KiB)

julia> @b collect(AbstractAlgebra.combinations(18,9))
3.293 ms (97269 allocs: 7.905 MiB)

julia> @b collect(combinations(18,9))
5.897 ms (145865 allocs: 7.790 MiB)
```

**Comment by thofma** [comment]
Added more tests and fixed the timings (usual inlining nonsense)

**Comment by thofma** [comment]
doctest failure is https://github.com/thofma/Hecke.jl/pull/1807, although I don't understand why this passed for example in https://github.com/thofma/Hecke.jl/actions/runs/13791087112/job/38570870779#step:6:150

**Comment by lgoettgens** [comment]
> doctest failure is [thofma/Hecke.jl#1807](https://github.com/thofma/Hecke.jl/pull/1807), although I don't understand why this passed for example in https://github.com/thofma/Hecke.jl/actions/runs/13791087112/job/38570870779#step:6:150

The failure is not in the documentation building job, but instead that is a separate job @benlorenz set up to catch exactly this type of issue.

**Comment by thofma** [comment]
ok thanks, I guess time to copy more things around

**Comment by joschmitt** [comment]
Just to say before anything becomes Official API: For the other functions of this kind (partitions, compositions, ...), we have a dedicated type `Partition`, `Composition`, ...

**Comment by thofma** [comment]
I am not quite sure what value that would add here.

**Comment by joschmitt** [comment]
> I am not quite sure what value that would add here.

We inherited the design from JuLie. From what I know, the idea was/is that this is helpful if one wants to work with combinations as objects themselves. So one can then write functions which take a combination as input or something.
I don't know whether there is interest in functionality in this direction, certainly not from my side.

**Comment by thofma** [comment]
OK, makes sense. I can add that.

**Comment by thofma** [comment]
Although, then it would not be a drop-in replacement for the current users of `AbstractAlgebra.combinations`

**Comment by lgoettgens** [comment]
> > Just to say before anything becomes Official API: For the other functions of this kind (partitions, compositions, ...), we have a dedicated type `Partition`, `Composition`, ...
>
> Although, then it would not be a drop-in replacement for the current users for `AbstractAlgebra.combinations`

What about having a dedicated return type for calls to `combinations(::Int, ::Int)`, as in these cases people want some combinatorial object. But calls to `combinations(::AbstractVector{T}, ::Int)` could still return an iterator over `Vector{T}`.
For the already existing functions of this kind, we only have the former variant, so it would be consistent in this regard. But the latter could still be used as a drop-in-replacement for `AbstractAlgebra.combinations`.

File: src/Combinatorics/EnumerativeCombinatorics/combinations.jl [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    state[i] += 1
+    if state[i] > (C.n - (C.k - i))
+      continue
+    end
+    for j in i+1:C.k
+      state[j] = state[j - 1] + 1
+    end
+    break
+  end
+  state[1] > C.n - C.k + 1 && return
```
Comment (fingolfin) [inline src/Combinatorics/EnumerativeCombinatorics/combinations.jl]: Maybe be explicit about returning nothing here, like in line 60?
```suggestion
  state[1] > C.n - C.k + 1 && return nothing
```

**Review by fingolfin — APPROVED** [review]
Looks nice

**Comment by lgoettgens** [comment]
Tagging triage to decide about the return type

**Comment by mjrodgers** [comment]
This looks good, and currently seems to be even a bit faster than the implementation used for `OrderedMultiIndexSet`.

The idea of having dedicated types is detailed [here](https://hackmd.io/Ef1t9wjbQwWlkrL4DQgYxQ).  Is there a reason it would break compatibility with the AbstractAlgebra version if it's done the same way as `Partition`? i.e. it is a subtype of `AbstractVector` and we have a `getindex` function so we can still access specific entries exactly as if it were a vector.

**Comment by fingolfin** [comment]
Discussed this in triage. My personal view: let's have this as `Oscar.combinations` and keep `AbstractAlgebra.combinations` separate (so that existing code using `combinations` can be easily adapted to use `AbstractAlgebra.combinations` for now, and can optionally change to `Oscar.combinations` later). 

Since we don't export `combinations` from Oscar right now, this would not be breaking for our users.

During triage nobody (dared to) disagree.

**Comment by thofma** [comment]
OK, I added a `Combination` type, which is just a thin wrapper around `Vector`:
```
julia> collect(combinations(['a', 'b', 'c'], 2))
3-element Vector{Combination{Char}}:
 ['a', 'b']
 ['a', 'c']
 ['b', 'c']
```
It might provide enough functionality to still be usable in the LieAlgebra module (instead of `Vector{Int}`). Tests will show.

**Comment by thofma** [comment]
@lgoettgens seems that the Lie algebra code is still happy with this change. Are you happy with it or should I restore the `LieAlgebras.combinations` thing?

**Comment by JohnAAbbott** [comment]
What guarantees does a `Combination` object make?  _e.g._ its elements are in strictly increasing order?
Can one have a `Combination` from a multi-set?  During triage I asked whether a `Combination` should also remember what it is a "subset" of; the general opinion seemed to be that it should not store this information.

**Comment by thofma** [comment]
> _e.g._ its elements are in strictly increasing order?

No, the elements have no ordering. What is ordered are the indices in the lexicographically ascending ordering. (We only consider combinations of `Vector`).

> Can one have a `Combination` from a multi-set?

No. The signature is `combinations(::Vector, ::Int)`.

> During triage I asked whether a `Combination` should also remember what it is a "subset" of; the general opinion seemed to be that it should not store this information.

Yes, this seems to be the general opinion.

**Comment by JohnAAbbott** [comment]
> No. The signature is `combinations(::Vector, ::Int)`.

Must the vector elements be distinct?  If not, what is the behaviour?  _e.g._ if the vector has length `n` then the combinations are just the vector elements obtained by indexing using the entries of a combination of the integers `1` up to `n`
Example: `collect(combinations([1,1,1],2)) == [Combination([1,1]), Combination([1,1]), Combination([1,1])]`

**Comment by thofma** [comment]
No, they don't have to be distinct and the example you give is the current behaviour. It is consistent with the other functions, e.g.,
```
julia> subsets([1, 1], 1)
2-element Vector{Vector{Int64}}:
 [1]
 [1]
```

**Comment by lgoettgens** [comment]
One usually doesn't want the function checking for uniqueness as that may have a severe performance impact (think e.g. of MPolyIdeals). If this is not the behavior you want to have, you can call `unique` explicitly beforehand

**Review by lgoettgens — APPROVED** [review]

**Comment by lgoettgens** [comment]
> @lgoettgens seems that the Lie algebra code is still happy with this change. Are you happy with it or should I restore the `LieAlgebras.combinations` thing?

This was exactly the change the code in `LieAlgebras/src/combinatorics.jl` was waiting for. I added (mostly naive) implementations there that should be removed/replaced once Oscar has something sensible for this.

Could you also please remove the corresponding testset from `experimental/LieAlgebras/test/Combinatorics-test.jl` or move them to your new testfile?

File: experimental/LieAlgebras/test/Combinatorics-test.jl [hunk 3]
```diff
@@ -1,99 +0,0 @@
...
-    @test collect(combinations(["1", "2", "3", "4"], 2)) ==
-      [["1", "2"], ["1", "3"], ["1", "4"], ["2", "3"], ["2", "4"], ["3", "4"]]
-    @test collect(combinations(["1", "2", "3", "4"], 3)) ==
-      [["1", "2", "3"], ["1", "2", "4"], ["1", "3", "4"], ["2", "3", "4"]]
-    @test collect(combinations(["1", "2", "3", "4"], 4)) == [["1", "2", "3", "4"]]
-    @test collect(combinations(["1", "2", "3", "4"], 5)) == String[]
-    @test collect(combinations(["1", "2", "3", "4"], 6)) == String[]
-  end
-
-  @testset "multicombinations" begin
```
Comment (lgoettgens) [inline experimental/LieAlgebras/test/Combinatorics-test.jl:20]: This testset and all below should be kept. Please just delete the `combinations` testset

**Review by lgoettgens — CHANGES_REQUESTED** [review]

**Review by lgoettgens — APPROVED** [review]
